### PR TITLE
Forbid writing to SD if write locked

### PR DESCRIPTION
**Pull request by windows-server-2003**
The bit 7 seems to be a SD card write lock flag(0 if locked).

### CODE REVIEW

**Comment by SirNapkin1334** [comment]
Interesting! I hadn't realized that you could write to the SD if it was locked, though it does make sense since the switch doesn't actually set anything on the inside of the SD, and there's no reason for the 3DS to include that functionality since it always has to write to the SD. But, apparently it does detect it, and ignore it? I can test it in about 9 hours.

**Comment by SirNapkin1334** [comment]
I checked it just now on my o3DSXL, and it worked. But you know this will be entirely useless on n3/2DS(XL), since microSD doesn't have a lock unless inserted into an adapter.
Good job, though!

**Comment by SirNapkin1334** [comment]
And, it functions as expected on the new systems (tested on n2DSXL), always allowing writes.

File: arm9/source/filesys/fsperm.c [hunk 1]
```diff
@@ -26,6 +26,12 @@ bool CheckWritePermissions(const char* path) {
     if ((drvtype & DRV_IMAGE) && !CheckWritePermissions(GetMountPath()))
         return false; // endless loop when mounted file inside image, but not possible
     
+    // SD card write lock check
+    if ((drvtype & (DRV_IMAGE | DRV_SDCARD | DRV_EMUNAND | DRV_ALIAS)) && SD_WRITE_LOCKED) {
```
Comment (d0k3) [inline arm9/source/filesys/fsperm.c]: Images can already have their origin on CTRNAND. You need to check the drive type of the image source. Besides, you don't need to check for images here. See the instrcuction block right above?

File: arm9/source/filesys/fsperm.h [hunk 2]
```diff
@@ -27,6 +27,8 @@
 #define PERM_YELLOW     (GetWritePermissions()&((PERM_SYS_LVL1&~PERM_SYS_LVL0)|(PERM_EMU_LVL1&~PERM_EMU_LVL0)|(PERM_SDDATA&~PERM_SDCARD)))
 #define PERM_GREEN      (GetWritePermissions()&(PERM_SDCARD|PERM_IMAGE|PERM_RAMDRIVE|PERM_EMU_LVL0|PERM_SYS_LVL0))
 
+#define SD_WRITE_LOCKED (!((*(volatile u16*)0x1000601C) & (0x1<<7)))
```
Comment (d0k3) [inline arm9/source/filesys/fsperm.h]: That really does not belong in here if it's checking an SD card specific register. It most likely belongs into sdmmc.h.

**Comment by d0k3** [comment]
@windows-server-2003 - just change the two things I remarked in the review, and I'll merge it.

**Comment by windows-server-2003** [comment]
I'll do it in 5 hours.

**Comment by d0k3** [comment]
Merged locally, thank you!

**Pull request closed without merging**